src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -4287,15 +4287,6 @@ module Queries = struct
         |> Runtime_config.to_yojson |> Yojson.Safe.to_basic )
 
   let fork_config =
-    let rec map_results ~f = function
-      | [] ->
-          Result.return []
-      | r :: rs ->
-          let open Result.Let_syntax in
-          let%bind r' = f r in
-          let%map rs = map_results ~f rs in
-          r' :: rs
-    in
     field "fork_config"
```

Comment:
This all happens within a single async cycle which pauses the daemon to both dump the accounts from the db and then json serialize them, both of which are slow operations. There are nearly 200k accounts on mainnet as of now. Are we sure that this is performant enough to not break a running daemon? Long async pauses are known to take a daemon offline and require a restart. I would need to see some concrete benchmarks of this function running against a mainnet ledger to be certain this is safe.

Reply:
I tried dumping the ledger this way multiple times on mainnet. It admittedly takes a couple of seconds, but the overall node's operations don't seem affected. I'll try gathering more concrete data, although I don't yet know how to go about it.

Reply:
Another approach would be to leverage `Async.In_thread` module to delegate the ledger extraction to a separate OS thread, so that GraphQL blocks on waiting until the thread is done, meanwhile letting other async computations to carry on. For this to work though, we need to ensure that no async computations take place in the OS thread, because there will be no Async machinery available to schedule them. This was the case before [2078969](https://github.com/MinaProtocol/mina/pull/14266/commits/20789695c1aa430f7f6c30cd0dea25fcc38218e6), but now it is no longer the case, as `Any_ledger.M.to_list` runs in `async`. So I'm inclided to actually revert that change, reverse the obtained list of accounts manually and then use `Async.In_thread` to move the entire computation to a separate OS thread. That should solve any performance concerns, right?

Reply:
I actually implemented the approach in [50992e7](https://github.com/MinaProtocol/mina/pull/14266/commits/50992e79902abf9b258934e4f0ae97570128e704). On mainnet, my node (on a relatively strong machine) takes about 40s. to export the ledger. With the latest commit included, the node continues to work normally during that time: it prodces logs and responds to status queries.

Reply:
`Async.In_thread` doesn't do what you think it does. The new thread holds the OCaml GC lock while it runs, and does not run concurrently to the existing thread. It's useful for enqueuing async kernel interactions in C code, but is not effective for running any concurrent OCaml code. Without multicore OCaml, there is no way to run OCaml code concurrently across different threads; at any one point in time, only 1 thread is allowed to lock the GC and run.

Reply:
With the latest commit ([c8c9a09](https://github.com/MinaProtocol/mina/pull/14266/commits/c8c9a09f00cc3957007d809e09ee6529ad859a38)) I added benchmarks for the most resource-consuming parts of the ledger export. Benchmarking the whole thing would require connecting to a real network or generating a ledger by hand – I decided both are too involved. So instead I extracted the ledger from the mainnet yesterday and used that as a basis to benchmark:
* serialization and deserialization of the runtime config
* converting accounts from ledger format to the config format.

Each of these tasks takes a couple seconds on my computer and the whole export takes about 40s. The node was responding to `mina client` commands as usual during this time. The config file I extracted is 400 MB large, so too big to attach it here. You can download it from [Google Cloud](https://drive.google.com/file/d/1TwbofVMl28s5eu6Wm7It3flETVtXfMbO/view?usp=drive_link).

Reply:
These are the results of the benchmark on my machine:
```
┌─────────────────────────────┬──────────┬──────────┬──────────┬──────────┬────────────┐
│ Name                        │ Time/Run │  mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├─────────────────────────────┼──────────┼──────────┼──────────┼──────────┼────────────┤
│ parse_runtime_config        │    3.08s │ 600.00Mw │ 166.90Mw │ 166.90Mw │     60.42% │
│ serialize_runtime_config    │    1.52s │ 133.36Mw │ 248.79Mw │  78.29Mw │     29.85% │
│ convert_accounts_for_config │    5.09s │ 462.53Mw │  14.82Mw │  14.82Mw │    100.00% │
└─────────────────────────────┴──────────┴──────────┴──────────┴──────────┴────────────┘
```
Note that `convert_accounts_for_config` gets executed two or three times for the ledger export. Unbenchmarked is the async function `Any_ledger.M.to_list`, but that's fine, because it probably blocks on I/O, which allows for yielding execution to non-graphql threads.

Reply:
@Sventimir `Any_ledger.M.to_list` is not async for rocksdb. None of the rocksdb I/O is nonblocking, it's all blocking.

Reply:
@Sventimir An async cycle of that length is in the range to take nodes offline, requiring a manual restart. The async scheduler and libp2p_helper do not interplay well when there is a long pause, and the node cannot recover. We need to fix this by adding manual scheduler yields. To do this, I would recommend yielding between each of the expensive steps. Additionally, I would recommend splitting up the `convert_accounts_for_config` step to yield after running batches of accounts, to get that to be less. It's not clear to me why `parse_runtime_config` takes so long, so I'm not sure what to do there right now.